master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java

```diff
@@ -139,10 +140,12 @@ public void updateBatchUnregisterShuffleMeta(List<String> shuffleKeys) {
     }
   }
 
-  public void updateAppHeartbeatMeta(String appId, long time, long totalWritten, long fileCount) {
+  public void updateAppHeartbeatMeta(
+      String appId, long time, long totalWritten, long fileCount, long shuffleFallbackCount) {
     appHeartbeatTime.put(appId, time);
     partitionTotalWritten.add(totalWritten);
     partitionTotalFileCount.add(fileCount);
+    shuffleTotalFallbackCount.add(shuffleTotalFallbackCount.sumThenReset() + shuffleFallbackCount);
```

Comment:
I am worried that the the LongAdder maintain an array for the added Long number. 

And it might be heavy if we always add element into it and does not make compaction.

Reply:
Can we make aggregation here?

https://github.com/apache/celeborn/blob/41fdb8ade10b97615533438607a2cb1add3a4232/master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java#L500

Reply:
The master will update the partitionSize periodically. 
https://github.com/apache/celeborn/blob/41fdb8ade10b97615533438607a2cb1add3a4232/master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala#L211-L223

Reply:
After checking the source code of JDK, I found your intention to call `sumThenReset` is invalid. This method won't clean the cells array.
```java
  public long sumThenReset() {
        Cell[] as = cells; Cell a;
        long sum = base;
        base = 0L;
        if (as != null) {
            for (int i = 0; i < as.length; ++i) {
                if ((a = as[i]) != null) {
                    sum += a.value;
                    a.value = 0L;
                }
            }
        }
        return sum;
    }
```

Reply:
According to the Java doc, I think the LongAdder handled the resizing of the cells arrays. Is there any evidence to tell that the cells array in LongAdder will increase without limitation?

```
/**
   * Handles cases of updates involving initialization, resizing,
   * creating new Cells, and/or contention. See above for
   * explanation. This method suffers the usual non-modularity
   * problems of optimistic retry code, relying on rechecked sets of
   * reads.
   *
   * @param x              the value
   * @param fn             the update function, or null for add (this convention
   *                       avoids the need for an extra field or function in LongAdder).
   * @param wasUncontended false if CAS failed before call
   */
  final void longAccumulate(long x, LongBinaryOperator fn,
                            boolean wasUncontended) {
```

Reply:
Ok, thanks for the info. I misunderstood the resizing.

There is a probe value for the thread that update the LongAdder, it is fine.
```
    /**
     * Returns the probe value for the current thread.
     * Duplicated from ThreadLocalRandom because of packaging restrictions.
     */
    static final int getProbe() {
        return UNSAFE.getInt(Thread.currentThread(), PROBE);
    }

```

Reply:
thanks a lot.